From 6749098f57b2849d3ca3ec5f8d04b916b6b0c8b3 Mon Sep 17 00:00:00 2001
From: Robby Cai <robby.cai@nxp.com>
Date: Thu, 11 Aug 2016 21:04:59 +0800
Subject: [PATCH 1126/5242] MLK-12881 media: camera: add error handing for
 BIT_RFF_OR_INT

commit  5d0e394d8f2d6e48c30ba5e4f3f2b92f56be2d4f from
https://source.codeaurora.org/external/imx/linux-imx.git

running following vte stress test will meet "mx6s-csi 21c4000.csi: mx6s_csi_irq_handler Rx fifo overflow"
and cannot be stopped to capture again.

i=0; while [ $i -lt 3000 ];do v4l2_capture_emma -D /dev/video1 -C 2 -M 0 -J 30,4 -W 640 -H 480;i=`expr $i + 1`;done

This patch adds the same handling as BIT_HRESP_ERR_INT for BIT_RFF_OR_INT
(RxFiFo OverFlow) to reset CSI as a recovery.

Signed-off-by: Robby Cai <robby.cai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/mxc/capture/mx6s_capture.c |    7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/media/platform/mxc/capture/mx6s_capture.c b/drivers/media/platform/mxc/capture/mx6s_capture.c
index d21fc4b..b546ccc 100644
--- a/drivers/media/platform/mxc/capture/mx6s_capture.c
+++ b/drivers/media/platform/mxc/capture/mx6s_capture.c
@@ -1058,8 +1058,11 @@ static irqreturn_t mx6s_csi_irq_handler(int irq, void *data)
 
 	if (status & BIT_RFF_OR_INT)
 		dev_warn(csi_dev->dev, "%s Rx fifo overflow\n", __func__);
+	if (status & BIT_HRESP_ERR_INT)
+		dev_warn(csi_dev->dev, "%s Hresponse error detected\n",
+			__func__);
 
-	if (status & BIT_HRESP_ERR_INT) {
+	if (status & (BIT_RFF_OR_INT|BIT_HRESP_ERR_INT)) {
 		/* software reset */
 
 		/* Disable csi  */
@@ -1084,8 +1087,6 @@ static irqreturn_t mx6s_csi_irq_handler(int irq, void *data)
 		/* Ensable csi  */
 		cr18 |= BIT_CSI_ENABLE;
 		csi_write(csi_dev, cr18, CSI_CSICR18);
-
-		pr_warning("Hresponse error is detected.\n");
 	}
 
 	if (status & BIT_ADDR_CH_ERR_INT) {
-- 
1.7.9.5

